[RFC] TensorCore 8bit implementations

XapaJIaMnu opened the pull request:
Hey @rhenry-nv ,

Thank you for your recent improvements to marian. Do you have any particular development goal with regards to marian?

 I am asking this, since I have been working on 8bit GPU version with and without tensor cores, using CUTLASS and I would like to avoid duplicated efforts where possible. In particular the GPU code for the 8bit GEMM is located here: https://github.com/XapaJIaMnu/marian-dev/blob/8bitgpu/src/tensors/gpu/integer_tools.cu

Do you have an opinion on the use of CUTLASS vs CUBLAS for TensorCore operations? Any particular comments on the GPU code?

Part of this pull request has been pending for review @marian-dev master, and the rest will incrementally go in, hopefully.

Cheers,

Nick

XapaJIaMnu committed:
Work around a compiler bug on gcc 5.4

XapaJIaMnu committed:
Remove commented-out code

XapaJIaMnu committed:
Update some comments and abort on bad matrix sizes.

XapaJIaMnu committed:
Restore proper hashing for shortlists

XapaJIaMnu committed:
Remove an old comment

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
fix the worse of the long term memory leakages

XapaJIaMnu committed:
Avoid consturcting extra nodes when working from preprocessed model

XapaJIaMnu committed:
Fix 16bit packed models

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Fix indentation and add comments

XapaJIaMnu committed:
Merge branch 'master' into intgemm_reintegrated

XapaJIaMnu committed:
Updated intgemm version

XapaJIaMnu committed:
Merge branch 'master' into intgemm_reintegrated

XapaJIaMnu committed:
initial shifted support. 0.1 BLEU

XapaJIaMnu committed:
Fix GPU compilation and remove some debugging

XapaJIaMnu committed:
Fix memoization

XapaJIaMnu committed:
Add a shiftedAll option with FakeBias

XapaJIaMnu committed:
Dump matrix stats

XapaJIaMnu committed:
Using static alphas

XapaJIaMnu committed:
Cached alphas working

XapaJIaMnu committed:
Prepare the output layer once

XapaJIaMnu committed:
Use cblas_sgemm_batched by default

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Optionally compress Wemb

XapaJIaMnu committed:
Switch the default to off

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Wrap the alphas in a node

XapaJIaMnu committed:
Use the correct size

XapaJIaMnu committed:
Update intgemm to master

XapaJIaMnu committed:
Update to latest intgemm

XapaJIaMnu committed:
Update intgemm

XapaJIaMnu committed:
wip minise mse

XapaJIaMnu committed:
Merge branch 'intgemm_reintegrated_computestats' of https://github.com/marian-nmt/marian-dev into intgemm_reintegrated_computestats

XapaJIaMnu committed:
Small fixes

XapaJIaMnu committed:
wip

XapaJIaMnu committed:
2*stddev -> 1.1*stddev

XapaJIaMnu committed:
some extra scripts

XapaJIaMnu committed:
update intgemm

XapaJIaMnu committed:
work with alrger models

XapaJIaMnu committed:
WiP fixing memory leaks

XapaJIaMnu committed:
hack that fixes the memory leak

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Some extra debug comments

XapaJIaMnu committed:
Update intgemm version and disable shortlist memoization

XapaJIaMnu committed:
Address pull request comments

XapaJIaMnu committed:
Midterm memory unnecessary

XapaJIaMnu committed:
Merge branch 'master' into intgemm_reintegrated

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
Updated changelog

XapaJIaMnu committed:
Update intgemm submodule

XapaJIaMnu committed:
Add extra comments

XapaJIaMnu committed:
Only compile intgemm when -DCOMPILE_CPU=ON

XapaJIaMnu committed:
Change command line parsing

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
initial 8bit gpu support

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
Merge remote-tracking branch 'origin/intgemm_reintegrated_computestats' into 8bitgpu

XapaJIaMnu committed:
Add CUTLASS as a submodule

XapaJIaMnu committed:
Use CUTLASS instead of CUBLAS for 8bit gemm

XapaJIaMnu committed:
Initial integer8 GPU structures

XapaJIaMnu committed:
Work around misaligned vectors when they occur

XapaJIaMnu committed:
Or -> and

XapaJIaMnu committed:
Merge branch 'intgemm_reintegrated_computestats' into 8bitgpu

XapaJIaMnu committed:
Make sure we always have aligned memory access

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
8bit dot and affine working

XapaJIaMnu committed:
Update Changelog and command line config parser

XapaJIaMnu committed:
Minor comment update

XapaJIaMnu committed:
Fix error due to merging

XapaJIaMnu committed:
Fix a command line parsing description due to merging

XapaJIaMnu committed:
update cmd parsing and WiP fixes for 16bit binary models

XapaJIaMnu committed:
Merge with the intgemm head

XapaJIaMnu committed:
Fixed loading binary models

XapaJIaMnu committed:
Merge branch 'intgemm_reintegrated_computestats' into 8bitgpu

XapaJIaMnu committed:
Make sure models with or without shortlist can use precomputed alphas

kpu committed:
Fast implementation of Select for most cases on CPU
Fixes #684

Measured:
enes.student.tiny11
xzcat sources.shuf.xz |head -n 10000
var (Cascade Lake) single core
based on intgemm_reintegrated_computestats branch

Before Total time: 66.69077s wall
After Total time: 61.20206s wall

kpu committed:
Changelog entry for Select

kpu committed:
Unreferenced parameter

kpu committed:
Unused lambda capture

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
Merge remote-tracking branch 'origin/kpu/fastselect' into intgemm_reintegrated_computestats

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
Fix aliases

kpu committed:
Optimize LayerNormalization with -ffast-math and lifting branches

kpu committed:
Update intgemm attempting to fix MSVC

kpu committed:
Merge branch 'master' into intgemm_reintegrated

kpu committed:
Microsoft isn't visual

kpu committed:
MSVC fixes

kpu committed:
Update intgemm with clearer MaxAbsolute code

kpu committed:
Use MSVC pragma syntax

kpu committed:
More MSVC patches

kpu committed:
CPUID support for MSVC

kpu committed:
MSVC: size_t and WordIndex already lose information

ugermann committed:
Update comments in marian_conv.cpp

Clarify use of saveGemmType (intgemm or fbgemm).

XapaJIaMnu committed:
Remove accidentally duplicated code

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev into master

XapaJIaMnu committed:
Partial CUDA11 Support

kpu committed:
Update intgemm, now compiling without error on MSVC

XapaJIaMnu committed:
Merge with master and add cuda11 support

XapaJIaMnu committed:
Enable compute75 when using cuda10 or newer and disable compute <50 when using CUDA11

XapaJIaMnu committed:
Update changelog

XapaJIaMnu committed:
Remove unnecessary bracket and unnecessary version check

XapaJIaMnu committed:
Merge branch 'compute75' into 8bitgpu

XapaJIaMnu committed:
Merge with upstream and add alphas to GPU, unfinished

XapaJIaMnu committed:
Initial alphas working

XapaJIaMnu committed:
Initial tensorcore implementation. 0 BLEU

XapaJIaMnu committed:
Updated cutlass to latest git master

XapaJIaMnu committed:
Its the B matrix that needs to be RowMajor'd

XapaJIaMnu committed:
Merge branch 'master' of https://github.com/marian-nmt/marian-dev into master

XapaJIaMnu committed:
Merge branch 'master' into intgemm_reintegrated

XapaJIaMnu committed:
Merge with master

XapaJIaMnu committed:
Small fix in command line parsing

XapaJIaMnu committed:
Fix unused parameter in windows build

XapaJIaMnu committed:
Ignore constant conditional in VisualC++

XapaJIaMnu committed:
Typo on the warning format for vs code

XapaJIaMnu committed:
More unused parameters for the windows build.

XapaJIaMnu committed:
Add bigobj to the windows build flags

XapaJIaMnu committed:
Tensor cores working for Dot (non-transposed)

XapaJIaMnu committed:
Tensorcores working for non-transposed matrices

XapaJIaMnu committed:
Transpose working with tensorcores

XapaJIaMnu committed:
Shortlist now working properly with 8bit GPU code

XapaJIaMnu committed:
Merge branch 'master' into intgemm_reintegrated

XapaJIaMnu committed:
Merge with intgemm_reintegrated

XapaJIaMnu committed:
Merge with upstream

XapaJIaMnu committed:
Add shared memory to performance critical kernel. Dubious effect

XapaJIaMnu committed:
Some more shared memory code

XapaJIaMnu committed:
UpdateIntgemmVersion

XapaJIaMnu committed:
Prepare for fused dequantize epilogue

XapaJIaMnu committed:
Merge with master

Commit:
Add CPU-only and GPU-only builds to Azure DevOps

Commit:
Move 'variables' to job level

Commit:
Add matrix.gpu to Ubuntu builds

Commit:
Try variables.var syntax for matrix variables

Commit:
Windows builds with marian-server

Commit:
Add comment on Boost on Azure-hosted runners

Commit:
Add matrix.static to Ubuntu builds

Commit:
Disable building marian-server in Windows CI

Commit:
Add Ubuntu build with minimum GCC and CMake supported

XapaJIaMnu committed:
Add fused dequantization GEMM

Commit:
Fix path to custom cmake

Commit:
Set CUDA_TOOLKIT_ROOT_DIR

Commit:
Use CUDA 10 in UbuntuBuildMinimal

Commit:
Remove call to marian-server.exe

XapaJIaMnu committed:
Small bugfixes

Commit:
Set timeout for Ubuntu jobs

Commit:
Build marian-server on Windows with Boost 1.72

Commit:
Clean up Ubuntu builds

XapaJIaMnu committed:
Fused bias addition GEMM

emjotde committed:
Merged PR 15320: Sync internal and external master

Updates internal master to external master.

Changes:
* Correct behavior for Pre-Norm transformer
* Small changes to CMake files

XapaJIaMnu committed:
How to tensorcore decoding

XapaJIaMnu committed:
Some additional info

Commit:
Merged PR 15332: Add CPU-only and GPU-only builds to Azure Pipelines

A few updates to Azure Pipelines:
* Adding CPU-only and GPU-only builds on Ubuntu
* Compiling Marian statically in some of the Ubuntu builds
* Ubuntu build with minimum supported versions of CMake (3.5.1), gcc (5.5), CUDA (10.0 due to GCC 5.5), no MKL
* Compiling marian-server with Boost 1.72 on Windows builds
* Minor clean up

Commit:
Merged PR 15561: Properly compile FBGEMM in CMake MSVC build

This fixes compilation of FBGEMM on Windows using CMake:
1. Compiling FBGEMM and cpuinfo statically
2. Forcing USE_STATIC_LIBS if USE_FBGEMM is set

XapaJIaMnu committed:
Avoid unnecessary transpose by using intgemm::PrepareBTransposed

XapaJIaMnu committed:
Fix compilation when doing a GPU only build

XapaJIaMnu committed:
Proper linking with intgemm

XapaJIaMnu committed:
Merge branch 'master' into intgemm_reintegrated

rhenry-nv committed:
Removed memcpy when batchA == batchB for batched matrix multiplication. Reduced the memcpys from 3 to 1 for the general case.

rhenry-nv committed:
Issue matrix multiplications in cublas and cusparse to the per thread default stream. This have issue each thread issue calls to their own stream instead of the global default stream when marian is compiled to use a default stream per thread

rhenry-nv committed:
Added new inference operator that peforms bias addition and optionally RELU for inference. When upgrading to cuda 11, the bias and relu can be fused into the matrix multiply with cublasLt.

rhenry-nv committed:
Batches logit retrieval from device to once memcpy is issued for all logits instead of one memcpy per logit.

rhenry-nv committed:
Reduced the number of times indices need to be copied to the GPU

rhenry-nv committed:
This commit slightly changes the output for certain models. It introduces a new operator to compute the sinusoidal embeddings on the GPU instead of doing CPU work. This is the reason for the differences since there are small differences in the float results for the positional embeddings. It also caches the triangle count mask for the transformer when doing inference to reduce device-hsot communication. This change should not affect the transofrmer output.

rhenry-nv committed:
Add licenses and change new class variable to have trailing underscore

rhenry-nv committed:
Introduces new operator that performs the bias addition, skip connection and layerNormailization for inference. Duplicates the layerNorm kernel since the LayerNormGrad kernel would no longer correspond to LayerNorm kernel if I extended the LayerNorm kernel.

rhenry-nv committed:
Update license in prod.cu

rhenry-nv committed:
update regression test repo

rhenry-nv committed:
Fix compilation for CPU only. Missed DISPATCH10 call when CUDA was undefined

XapaJIaMnu committed:
update to latest cutlass

rhenry-nv committed:
Changed getherFromIndices to use pointers instead of get and set for CPU backend. Should be more performant + the code is more readable.

rhenry-nv committed:
Fix compile errors on Windows and MacOS.

XapaJIaMnu committed:
Merge with master and intgemm_reintegrated

XapaJIaMnu committed:
Add bac fastselect implementation

XapaJIaMnu committed:
Merge with nvidia gpu optimisation. Bug. Bleu is lower than expected

XapaJIaMnu committed:
fix intgemm BLEU

XapaJIaMnu committed:
Cpuonly build working

XapaJIaMnu committed:
GPU only build working

XapaJIaMnu committed:
Fix int8Alpha decoding

rhenry-nv committed:
Add synchronize before free in gpu/prod.cpp

rhenry-nv committed:
Adds a cache by name to the graph. Moves the lemmaHasFactorGroup tensor to the GPU. Implements an inference operator to mask out the lemmas of a factored vocab on GPU.

XapaJIaMnu committed:
Compute alphas on the GPU. Still low BLEU

rhenry-nv committed:
Fixes reductions into scalars for <= 32 input elements. Only affects reductions where 0 is not the identity

rhenry-nv committed:
Update CHANGELOG.md

rhenry-nv committed:
Adds cub as a submodule

XapaJIaMnu committed:
Fix CPU only compilation

rhenry-nv committed:
Redo topk implementation

rhenry-nv committed:
Fix compile error on windows

rhenry-nv committed:
Adds bigobj flag to windows builds

XapaJIaMnu committed:
Fix alphas on GPU and remove debugging info

XapaJIaMnu committed:
Merge latest nvidia optimisations

XapaJIaMnu committed:
Merge latest nvidia fixes

XapaJIaMnu committed:
Cuda 11 Support for nvidia's improvements

XapaJIaMnu committed:
Cuda 11 Support for nvidia's improvements

rhenry-nv committed:
Fixes a race condition in gAddFactorMaxes. Reworks API to make this race more obvious in future

rhenry-nv commented:
Hi Nikolay,

I am mainly trying to improve inference performance by mainly looking for low hanging fruit. I did not plan to add int8 support but I am happy to see that this is under development!

Re cublas/cutlass - Are you using any special int8 formats to do the gemm?  At a high level I would maybe stick to cublas if you want to just do GEMMs with int8 in row/col major. One reason to prefer cutlass is that it allows one to fuse the dequantization and other post processing ops into the GEMM which can reduce some of the overhead associated with quantizing. If you're using the special int8 format then I would nudge in favor of cutlass for its fusion so you hide overheads.

I can take a look at the file you linked and provide feedback in the next few days.

rhenry-nv committed:
Remove redundant includes in topk

XapaJIaMnu commented:
@rhenry-nv  and tagging @kpu here.

We are indeed fusing the dequantisation and bias addition into the GEMM, which is why we looked at CUTLASS. I am yet to fuse the activation functions but I will do that as well. 

Before you did your improvements in marian, this codepath yielded about 20% performance improvement from fp32, but afterwards fp32 is slightly faster. One of the problems i have with CUTLASS is that it has a million parameters to tune , and in addition to that, I am not sure the performance critical quantisation kernel is written in the most optimal way possible (even though it's just a simple streaming operation).

rhenry-nv commented:
Hi Nikolay,

I briefly skimmed the code but I have one question. Why the use of cudaMallocManaged? Would it be possible to use the Marian allocator instead?

XapaJIaMnu commented:
@rhenry-nv the two places where `sudaMallocManaged` is used are not in a performance critical path: 

They are only run when the quantization multipliers for the activations are precomputed and then baked into the .npz so that we don't have to compute them every time during decoding. But yeah, it would be possible to trickle down memory allocation from Marian's allocator, but since this is not used during decoding anyways, I didn't bother.

The only kernel (other than the GEMMs) that is in the performance critical codepath (assuming fused dequantization/bias addition) is https://github.com/XapaJIaMnu/marian-dev/blob/8bitgpu/src/tensors/gpu/integer_tools.cu#L397

Thanks for looking at it:

Nick

rhenry-nv committed:
Fixes topk when all elements are the lowest value of the input type

rhenry-nv committed:
Adds license to nth_element.cu

XapaJIaMnu committed:
Merge with latest nvidia optimisations

XapaJIaMnu committed:
preliminary fused RELU support, disabled due to bug with CUTLASS

rhenry-nv committed:
Removes unneccessary cub include outside of header guards

rhenry-nv committed:
Fixes bug for certain reductions (#746)

* Fixes reductions into scalars for <= 32 input elements. Only affects reductions where 0 is not the identity
* Update CHANGELOG.md
* Adds space before "?"
* Adds comment explaining increase in margin for reduction tests. Adds axis comment to argument to reduce functions. Adds more tests for small reduction operators

rhenry-nv committed:
Handles not found in topk stage 2

rhenry-nv committed:
Removes unneeded include

XapaJIaMnu committed:
Fix MPI on GCC8+ (#742)

ykim362 committed:
Move batched GEMM back (#710)

emjotde committed:
Merged PR 15896: Add --after N option to supersede --after-batches and --after-epochs

Replace `--after-batches N` and `--after-epochs N` with `--after Nu/Ne` which allows to specify updates, epochs, target labels with units, e.g.:
* `--after 30Gt` or `--after 50ku` or `--after 10e`
* Can also combine multiple criteria: `--after 30Gt,50ku,10e` and will stop when whichever hits first

Changes default `cost-type` from `ce-mean` to `ce-sum` and turns `display-label-counts` on by default.

rhenry-nv committed:
Adds fp16 support for AddFactorMaxes

emjotde committed:
Merged PR 16144: Merge Cross-Entropy with Label-Smoothing operation

* Compute label-smoothing within the cross-entropy node, should result in faster training.

emjotde committed:
Merged PR 16162: ChrF validation metric

This PR adds ChrF as a validation metric. This follows the implementation from SacreBLEU.

emjotde committed:
Merged PR 16168: Add redefinition of logical epoch

Adds e.g. --logical-epoch 1Gt (or other units) that alters the way the epoch counter is displayed. The actual underlying counter in form data passes is not changed. This is essentially a logging change that will now display the epoch as a fractional multiple of the chosen unit.

Example for `--logical-epoch 100Mt`:
```
[2020-11-02 04:14:16] Ep. 4.8602 : Up. 16755 : Sen. 1,088,000 : Cost 1.17630422 * 1,993,304 @ 31,985 after 486,015,051 : Time 61.36s : 32483.55 words/s
[2020-11-02 04:15:18] Ep. 4.8803 : Up. 16825 : Sen. 1,162,648 : Cost 1.17474616 * 2,009,996 @ 37,740 after 488,025,047 : Time 61.88s : 32480.17 words/s
[2020-11-02 04:16:19] Ep. 4.9002 : Up. 16893 : Sen. 1,235,200 : Cost 1.17799997 * 1,990,844 @ 26,173 after 490,015,891 : Time 60.47s : 32920.16 words/s

emjotde committed:
merge with internal

rhenry-nv committed:
Patch for multiRows to work with fp16

rhenry-nv committed:
Implement new features from cuda 11. Implements spmm using new cusparse call instead of deprecated gemmi. Implements cublasLt affine operation with optional relu fusion. Not fully tested.

emjotde committed:
Merged PR 16219: Allow to set epoch display width

Allow to set the display width of the fractional pert of a logical epoch.

afaji committed:
Quantized model finetuning  (#690)

* enable quantized training

delong-coder committed:
Add Triton Marian backend running in AzureML Inference Environment (#749)

* Add Triton Marian backend running in AzureML Inference Environment

emjotde committed:
Merge branch 'master' into pmaster

rhenry-nv committed:
Updates CUDA 11 affine support and SpMM support. CublasLt has a bug with is causes a misaligned read when the bias is not a multiple of 8 (this occurs during slicing). This case is handled specially. This bug will be fixed in a subsequent release of cuda.

rhenry-nv committed:
Adds guard for cuda 11 around spmm call

rhenry-nv committed:
Adds check to cublas and does reductions in AddFactorMaxes in parallel. Fixes buf in AddFactorMaxes

rhenry-nv committed:
Changes alignment requirement back to 8

emjotde committed:
Merged PR 15925: Training embedder separation with margin losses

* This PR adds training of embedding spaces with better separation based on https://arxiv.org/abs/2007.01852
* We can now train with in-batch negative examples or a handful of hand-constructed negative examples provided in a tsv-file.

afaji committed:
Hotfix: crash when --quantize-optimization-steps is set to 0 (#759)

* fix problem if the optimization step is set to 0
* set the first error residual to 0

emjotde committed:
Update sentencepiece to newest version (#753)

- Updates sentencepiece to the newest version (removes dependency on protobuf)
- Enable SentencePiece compilation by default since there is no dependency in protobuf anymore.

emjotde committed:
update regression test pointer

emjotde committed:
Merged PR 16294: Change stopping criterion for mini-batch-fit binary search

This PR changes the stopping criterion for mini-batch-fit binary search to better maximize batch size.

emjotde committed:
Merged PR 16337: Update sentencepiece to new version

This updates the SentencePiece version in Marian to a much more recent revision. Due to that there is no dependency on Protobuf anymore.

emjotde committed:
merge with internal master

emjotde committed:
update version based on number of PRs from last time

afaji committed:
Fix the allocation size for opt-step Tensor in the quantizer (#761)

emjotde committed:
Fixes #763

emjotde committed:
Merge branch 'master' of github.com:marian-nmt/marian-dev into pmaster

snukky committed:
Update Boost in GitHub checks to 1.72 (#764)

snukky committed:
Remove ::set-env from GitHub checks for Windows (#766)

XapaJIaMnu committed:
Merge remote-tracking branch 'nvidia/gpu_optimizations' into 8bitgpu

XapaJIaMnu committed:
Merge with upstream master and update cutlass

XapaJIaMnu committed:
RE-enable fused relu with the latest CUTLASS code

XapaJIaMnu committed:
Fix CUTLASS build flags

XapaJIaMnu committed:
Enable the use of apple accelerate framework on macs and turn it on by default

XapaJIaMnu committed:
Ampere RTX support

XapaJIaMnu committed:
Updated cutlass

XapaJIaMnu committed:
Add Sm80 support (not working yet)

XapaJIaMnu committed:
SM80 fix for 3090tis

Commit:
8bitgpu fp16 (#3)

* attempt to enable cutlass tensorcore with fp16: compilation OK

* FP16 support for NodeOp (partial implementation)

* switch to reinterpret_cast

* add cutlass FP16 support from Nick

* add cutlass FP16 support for DotNodeOp

* set quant NodeOp type

* more NodeOp changes for FP16 support

* remove debugging info

* some comments and aborts

XapaJIaMnu committed:
backport add_all.inc from upstream

This is done so that we can compile with newer cuda versions. It also adds some extra templates that are not used in this branch of marian, but it doesn't impede translation or compilation.
